runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java

```diff
@@ -146,22 +154,92 @@ private BundleProcessor(
+    private final CompletionStage<BeamFnApi.ProcessBundleResponse> response;
+    private final CloseableFnDataReceiver<WindowedValue<InputT>> inputReceiver;
+    private final Map<BeamFnApi.Target, InboundDataClient> outputClients;
 
-    public abstract CloseableFnDataReceiver<WindowedValue<InputT>> getInputReceiver();
-    public abstract Map<BeamFnApi.Target, InboundDataClient> getOutputClients();
-
-    public static <InputT> ActiveBundle<InputT> create(
+    private ActiveBundle(
         String bundleId,
         CompletionStage<BeamFnApi.ProcessBundleResponse> response,
-        CloseableFnDataReceiver<WindowedValue<InputT>> dataReceiver,
+        CloseableFnDataReceiver<WindowedValue<InputT>> inputReceiver,
         Map<BeamFnApi.Target, InboundDataClient> outputClients) {
-      return new AutoValue_SdkHarnessClient_ActiveBundle<>(
-          bundleId, response, dataReceiver, outputClients);
+      this.bundleId = bundleId;
+      this.response = response;
+      this.inputReceiver = inputReceiver;
+      this.outputClients = outputClients;
+    }
+
+    /**
+     * Returns an id used to represent this bundle.
+     */
+    public String getBundleId() {
+      return bundleId;
+    }
+
+    /**
+     * Returns a {@link FnDataReceiver receiver} which consumes input elements forwarding them
+     * to the SDK. When
+     */
+    public FnDataReceiver<WindowedValue<InputT>> getInputReceiver() {
+      return inputReceiver;
+    }
+
+    /**
+     * Blocks till bundle processing is finished. This is comprised of:
+     * <ul>
+     *   <li>closing the {@link #getInputReceiver() input receiver}.</li>
+     *   <li>waiting for the SDK to say that processing the bundle is finished.</li>
+     *   <li>waiting for all inbound data clients to complete</li>
+     * </ul>
+     *
+     * <p>This method will throw an exception if bundle processing has failed.
+     * {@link Throwable#getSuppressed()} will return all the reasons as to why processing has
+     * failed.
+     */
+    @Override
+    public void close() throws Exception {
+      Exception exception = null;
+      try {
+        inputReceiver.close();
+      } catch (Exception e) {
+        exception = e;
+      }
+      try {
+        // We don't have to worry about the completion stage.
+        if (exception == null) {
+          MoreFutures.get(response);
+        } else {
+          // TODO: Handle aborting the bundle being processed.
+          throw new IllegalStateException("Processing bundle failed, TODO: abort bundle.");
+        }
+      } catch (Exception e) {
+        if (exception == null) {
+          exception = e;
+        } else {
+          exception.addSuppressed(e);
+        }
+      }
+      for (InboundDataClient outputClient : outputClients.values()) {
+        try {
+          // If we failed processing this bundle, we should cancel all inbound data.
+          if (exception == null) {
+            outputClient.awaitCompletion();
```

Comment:
Could the failure of a single inbound client cause other clients to fail? If so, would this be captured by the overall bundle result? If not, you will need to wait for completion of the first-finished result rather than an arbitrary client here.

Reply:
We await completion on all outbound clients.

No, the bundle result represents to the best knowledge that the SDK did everything it needed to. There can still be wire transfer/network/decoding failures that the SDK wouldn't be aware of after sending bundle process completion.